File: main/src/main/java/lavaplayer/manager/AudioConfiguration.java
package lavaplayer.manager;

import lavaplayer.format.AudioDataFormat;
import lavaplayer.format.StandardAudioDataFormats;
import lavaplayer.track.playback.AllocatingAudioFrameBuffer;
import lavaplayer.track.playback.AudioFrameBufferFactory;

/**
 * Configuration for audio processing.
 */
public class AudioConfiguration {
    public static final int OPUS_QUALITY_MAX = 10;

    private volatile ResamplingQuality resamplingQuality;
    private volatile int opusEncodingQuality;
    private volatile AudioDataFormat outputFormat;
    private volatile boolean filterHotSwapEnabled;
    private volatile AudioFrameBufferFactory frameBufferFactory;

    /**
     * Create a new configuration with default values.
     */
    public AudioConfiguration() {
        resamplingQuality = ResamplingQuality.LOW;
        opusEncodingQuality = OPUS_QUALITY_MAX;
        outputFormat = StandardAudioDataFormats.DISCORD_OPUS;
        filterHotSwapEnabled = false;
        frameBufferFactory = AllocatingAudioFrameBuffer::new;
    }

    public ResamplingQuality getResamplingQuality() {
        return resamplingQuality;
    }

    public void setResamplingQuality(ResamplingQuality resamplingQuality) {
        this.resamplingQuality = resamplingQuality;
    }

    public int getOpusEncodingQuality() {
        return opusEncodingQuality;
    }

    public void setOpusEncodingQuality(int opusEncodingQuality) {
        this.opusEncodingQuality = Math.max(0, Math.min(opusEncodingQuality, OPUS_QUALITY_MAX));
    }

    public AudioDataFormat getOutputFormat() {
        return outputFormat;
    }

    public void setOutputFormat(AudioDataFormat outputFormat) {
        this.outputFormat = outputFormat;
    }

    public boolean isFilterHotSwapEnabled() {
        return filterHotSwapEnabled;
    }

    public void setFilterHotSwapEnabled(boolean filterHotSwapEnabled) {
        this.filterHotSwapEnabled = filterHotSwapEnabled;
    }

    public AudioFrameBufferFactory getFrameBufferFactory() {
        return frameBufferFactory;
    }

    public void setFrameBufferFactory(AudioFrameBufferFactory frameBufferFactory) {
        this.frameBufferFactory = frameBufferFactory;
    }

    /**
     * @return A copy of this configuration.
     */
    public AudioConfiguration copy() {
        AudioConfiguration copy = new AudioConfiguration();
        copy.setResamplingQuality(resamplingQuality);
        copy.setOpusEncodingQuality(opusEncodingQuality);
        copy.setOutputFormat(outputFormat);
        copy.setFilterHotSwapEnabled(filterHotSwapEnabled);
        copy.setFrameBufferFactory(frameBufferFactory);
        return copy;
    }

    /**
     * Resampling quality levels
     */
    public enum ResamplingQuality {
        HIGH,
        MEDIUM,
        LOW
    }
}

File: main/src/main/java/lavaplayer/manager/AudioConfiguration.kt
package lavaplayer.manager

import lavaplayer.filter.ResamplingQuality
import lavaplayer.format.AudioDataFormat
import lavaplayer.format.StandardAudioDataFormats
import lavaplayer.track.playback.AllocatingAudioFrameBuffer
import lavaplayer.track.playback.AudioFrameBufferFactory

/**
 * Audio processing configuration.
 */
class AudioConfiguration {
    companion object {
        const val OPUS_QUALITY_MAX = 10
    }

    @get:JvmName("isFilterHotSwapEnabled")
    var filterHotSwapEnabled: Boolean = false
    var resamplingQuality: ResamplingQuality = ResamplingQuality.LOW
    var outputFormat: AudioDataFormat = StandardAudioDataFormats.DISCORD_OPUS
    var opusEncodingQuality: Int = OPUS_QUALITY_MAX
        set(value) {
            field = value.coerceIn(0..OPUS_QUALITY_MAX)
        }

    lateinit var frameBufferFactory: AudioFrameBufferFactory

    init {
        useFrameBufferFactory(::AllocatingAudioFrameBuffer)
    }

    operator fun invoke(block: AudioConfiguration.() -> Unit): AudioConfiguration {
        return apply(block)
    }

    fun useFrameBufferFactory(factory: AudioFrameBufferFactory): AudioConfiguration {
        frameBufferFactory = factory
        return this
    }

    fun clone(): AudioConfiguration {
        val copy = AudioConfiguration()
        copy.resamplingQuality = resamplingQuality
        copy.opusEncodingQuality = opusEncodingQuality
        copy.outputFormat = outputFormat
        copy.filterHotSwapEnabled = filterHotSwapEnabled
        copy.frameBufferFactory = frameBufferFactory
        return copy
    }
}
